### PR TITLE
mgr/nfs: include pseudo in JSON output when nfs export apply -i fails

### PR DESCRIPTION
**Pull request by dparmar18**
```
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
    "state": "error"
  },
 {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.2'",
    "state": "error"
 }
]
```

it is a bit of hassle to know which export's update failed and the dict contains no info on what was the errno, therefore with something like below, it is much better to figure out which export block needs correction

```
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
    "pseudo": "cephfs1",
    "state": "error"
  },
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.2'",
    "pseudo": "cephfs2",
    "state": "error"
  }
]
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by dparmar18** [comment]
with this it will be easy to verify errno in our test suite too

**Review by phlogistonjohn — APPROVED** [review]
Works for me, but you might want to negate (or abs value) the errno so it's always positive. The negative errno is an internal implmeentation thing I think ceph borrowed from the linux kernel (or likewise), and doesn't need to be exposed to the caller. But it's not a biggie either way.

**Comment by dparmar18** [comment]
> Works for me, but you might want to negate (or abs value) the errno so it's always positive. The negative errno is an internal implmeentation thing I think ceph borrowed from the linux kernel (or likewise), and doesn't need to be exposed to the caller. But it's not a biggie either way.

Makes sense, will push the change.

**Comment by dparmar18** [comment]
Other thing I'd like to change is sending EIO, I think EINVAL is a better fit since it indicates the export block contains invalid entries and needs correction.

**Comment by dparmar18** [comment]
other than that, the only thing pending is a test case that validates these changes with multiple export blocks(this will also add some code coverage since we don't have any test cases testing multiple exports update)

**Comment by dparmar18** [comment]
last push fixed a bug where if a non NFSException occurred then it migh've crashed with `no attribute 'errno'` since exceptions derived from Exception class don't have `errno`

**Comment by dparmar18** [comment]
@phlogistonjohn one more thing i was trying to fix was getting

`Error EIO:` at the end of the response. It looks incomplete. I was thinking about changing it to `Error EINVAL: unknown error encountered.`

**Comment by phlogistonjohn** [comment]
> Other thing I'd like to change is sending EIO, I think EINVAL is a better fit since it indicates the export block contains invalid entries and needs correction.

I'm not sure I 100% agree but I don't feel strongly about it and won't object. Go ahead with that change.

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn one more thing i was trying to fix was getting
> 
> ```
> Error EIO:
> ```
> 
> at the end of the response. It looks incomplete. I was thinking about changing it to
> 
> ```
> Error EINVAL: unknown error encountered.
> ```

Works for me, but you'll probably need to make changes to object_format.py to do so. Due to that added complexity you might want to skip it for now. Otherwise, I can be avaiable to help you with adding the necessary hooks to object_format to have arbitrary non-exception types also provide a status value.

**Comment by dparmar18** [comment]
> > @phlogistonjohn one more thing i was trying to fix was getting
> > ```
> > Error EIO:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > at the end of the response. It looks incomplete. I was thinking about changing it to
> > ```
> > Error EINVAL: unknown error encountered.
> > ```
> 
> Works for me, but you'll probably need to make changes to object_format.py to do so. Due to that added complexity you might want to skip it for now. Otherwise, I can be avaiable to help you with adding the necessary hooks to object_format to have arbitrary non-exception types also provide a status value.

Got it. Let's keep it for another PR, I'll raise a tracker

**Comment by dparmar18** [comment]
> > Works for me, but you'll probably need to make changes to object_format.py to do so. Due to that added complexity you might want to skip it for now. Otherwise, I can be avaiable to help you with adding the necessary hooks to object_format to have arbitrary non-exception types also provide a status value.
> 
> Got it. Let's keep it for another PR, I'll raise a tracker

@phlogistonjohn https://tracker.ceph.com/issues/62641

**Comment by vshankar** [comment]
> ```
> [
>   {
>     "errno": 22,
>     "export_id": 1,
>     "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
>     "state": "error"
>   },
>   {
>     "errno": 22,
>     "export_id": 2,
>     "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.2'",
>     "state": "error"
>   }
> ]
> ```

With the errno made specific to the export-id as as part of the failed message response, the -EIO becomes a dummy errno that's used to indicate just the command failure. It seems a bit off-putting to me when I looked at it initially. That said including the export-id in the failure message is adding value, so +1 for that, but I'm not so sure about having the errno in there.

**Comment by dparmar18** [comment]
> > ```
> > [
> >   {
> >     "errno": 22,
> >     "export_id": 1,
> >     "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
> >     "state": "error"
> >   },
> >   {
> >     "errno": 22,
> >     "export_id": 2,
> >     "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.2'",
> >     "state": "error"
> >   }
> > ]
> > ```
> 
> With the errno made specific to the export-id as as part of the failed message response, the -EIO becomes a dummy errno that's used to indicate just the command failure. It seems a bit off-putting to me when I looked at it initially. That said including the export-id in the failure message is adding value, so +1 for that, but I'm not so sure about having the errno in there.

In case of a single export update failure it definitely would be confusing for the user (and even more confusing after the code to display respective errno is pushed). In case of multiple failures, I guess the msg is more than enough to help user understand the issue. I'm now too in favour of removing errno from the dict.

File: src/pybind/mgr/nfs/export.py
Comment (vshankar) [inline src/pybind/mgr/nfs/export.py:565]: Why is the `state` field useful?

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py:565]: the -EIO errno is telling that (some) exports failed updation - isn't that enough?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: > the -EIO errno is telling that (some) exports failed updation - isn't that enough?

found it, it has a utility in `class AppliedExportResults` where the `status` is used to return EIO

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: > Why is the `state` field useful?

I've get an idea, we can get rid of this `state` by changing `if value.get("state", "") == "error"` to `if value.pop("state", "") == "error"`in 

https://github.com/ceph/ceph/blob/9d7c18257836dc888b4a300f3b5af9f080910986/src/pybind/mgr/nfs/export.py#L171-L173

this way it would just be:
```
[
  {
    "export_id": 1,
    "msg": "Failed to apply export: filesystem foo not found",
  },
  {
    "export_id": 2,
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.2'",
  }
]
```
and this looks really neat

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: @vshankar @phlogistonjohn what do you think?

Comment (phlogistonjohn) [reply src/pybind/mgr/nfs/export.py:565]: I think there should be a clear indicator for each item showing if it was ok or not, for automation code that might be consuming the JSON. Code needs to easily distinguish between errors and non-errors. That could be via state field or via always having an 'errno' field on error items. But requiring code to scrape text is not a good approach and defeats the point of using JSON output.

**Comment by dparmar18** [comment]
passed http://pulpito.front.sepia.ceph.com/dparmar-2023-09-04_12:32:42-orch:cephadm-return_errno_and_export_id_at_export_update_failure-distro-default-smithi/

File: src/pybind/mgr/nfs/export.py
Comment (vshankar) [reply src/pybind/mgr/nfs/export.py:565]: @phlogistonjohn Agree. So we agree to keep the `state` field and possible values for those are `success` or `error`.

**Comment by dparmar18** [comment]
@vshankar ^^

File: src/pybind/mgr/nfs/export.py
Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: @vshankar possible values for `state` can be `added`/`updated`/`error`

(FYI `ceph nfs export apply -i <{json/conf}_file>`) not only can update an export but create a new export too.

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py:565]: Fair enough. Although `updated` it pretty much depicting the same as `added`, I'm ok to separate those out. 👍

**Comment by vshankar** [comment]
@dparmar18 please remove commit bcdd00c65fafdf34bd8f7ddd3ff78460c17126b0 as PR #52924 is ready to be merged.

**Comment by dparmar18** [comment]
> @dparmar18 please remove commit [bcdd00c](https://github.com/ceph/ceph/commit/bcdd00c65fafdf34bd8f7ddd3ff78460c17126b0) as PR #52924 is ready to be merged.

@vshankar done

**Comment by dparmar18** [comment]
let us also get review from rgw folks before since merging since this also affects RGW exports

**Comment by dparmar18** [comment]
last push made no changes to the code, just some minor corrections in the commit messages.

**Comment by dparmar18** [comment]
last push: rebase

File: qa/tasks/cephfs/test_nfs.py
Comment (vshankar) [inline qa/tasks/cephfs/test_nfs.py:928]: nit: typo - `export_id`.

File: src/pybind/mgr/nfs/export.py
Comment (vshankar) [reply src/pybind/mgr/nfs/export.py:565]: Do we have tests that check the `state` field?

File: qa/tasks/cephfs/test_nfs.py
Comment (vshankar) [inline qa/tasks/cephfs/test_nfs.py:768]: This bit won't be required once PR #53240 is merged, yes?

**Review by vshankar — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_nfs.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:768]: that PR is going to be closed. it is going to be resolved in the underlying code(object_format.py)

Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:768]: In that case, please close the PR. anything blocking?

**Comment by dparmar18** [comment]
https://github.com/ceph/ceph/pull/53177#discussion_r1318169990

i dont know why i can't reply to this comment so quoting its link here

no, we dont have any tests for it

File: qa/tasks/cephfs/test_nfs.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py]: done

File: qa/tasks/cephfs/test_nfs.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py]: fixed

File: qa/tasks/cephfs/test_nfs.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:768]: no, it just completely went off my mind to do it

**Comment by dparmar18** [comment]
last changes went fine http://pulpito.front.sepia.ceph.com/dparmar-2023-09-08_12:16:38-orch:cephadm-return_errno_and_export_id_at_export_update_failure-distro-default-smithi/

**Comment by vshankar** [comment]
> last changes went fine http://pulpito.front.sepia.ceph.com/dparmar-2023-09-08_12:16:38-orch:cephadm-return_errno_and_export_id_at_export_update_failure-distro-default-smithi/

Sure. But have you looked at comment https://github.com/ceph/ceph/pull/53177#discussion_r1318169990 ?

**Comment by dparmar18** [comment]
> > last changes went fine http://pulpito.front.sepia.ceph.com/dparmar-2023-09-08_12:16:38-orch:cephadm-return_errno_and_export_id_at_export_update_failure-distro-default-smithi/
> 
> Sure. But have you looked at comment [#53177 (comment)](https://github.com/ceph/ceph/pull/53177#discussion_r1318169990) ?

https://github.com/ceph/ceph/pull/53177#discussion_r1319710795

File: src/pybind/mgr/nfs/export.py
Comment (vshankar) [reply src/pybind/mgr/nfs/export.py:565]: It makes sense to add the changes here (state+test) IMO.

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: the tracker is of adding export_id to err dict, i dont think the test case is relevant to the tracker. the test case will be relevant in the tracker where i'll write the code to fix the errno and the err string, where multiple blocks will be supplied and the response could be added/error.

**Comment by dparmar18** [comment]
removed [changes](https://github.com/ceph/ceph/compare/78091f76986f5a8801f5376cd89aa45b4997e6aa..be073ec01a5f31c53730934e35fa940102bddb0a) made to test case `update_with_invalid_values` since https://github.com/ceph/ceph/pull/53431 will deal with it and with single export update failure, the actual errno (instead of EIO) will be sent to CLI

File: src/pybind/mgr/nfs/export.py
Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py:565]: will be added in https://github.com/ceph/ceph/pull/53431

**Comment by dparmar18** [comment]
@vshankar ping

**Comment by dparmar18** [comment]
[last push](https://github.com/ceph/ceph/compare/be073ec01a5f31c53730934e35fa940102bddb0a..bf5b9d43e027817891dade87c92bb03e9ee545b8) was for the test case(no code changes) - made the test case name more relevant + added doc string + enhanced a comment

File: qa/tasks/cephfs/test_nfs.py
Comment (vshankar) [inline qa/tasks/cephfs/test_nfs.py:935]: How is this useful to the user who is passing a export block to create a new export, which fails, but the export_id in present in the output json.

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: it might not be a big deal if there is a single block or a few blocks but if there are extensive exports then it will helps figure out which export block creation/update failed

Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:935]: Sure, that's fine for existing exports, but for exports which are new and are failing creation, the export_id isn't of any use, right?

Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:935]: ... and following up on this: On the next call if the export creation is successful, what export_id is assigned?

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: if it's a new export creation request, a call is made:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L780-L784

if you see, before processing anything in `create_export_from_dict()`, it provides an ID to the export, if everything goes fine, the code block below is executed which saves the new export to rados else the exception is raised and thus nothing is committed anywhere:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L786-L789

this means when the export creation failed, we did have the export_id of the new export that was to be created, so by having the export_id in the JSON output (of the export that failed to create) we mean to say "hey you wanted to create an export but didn't provide an export id, that's fine we gave it X which was the next available integer but due to something in the export block, it(X) could not be created".

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: > ... and following up on this: On the next call if the export creation is successful, what export_id is assigned?

the code that finds the next available ID is:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L282-L290

i.e. it checks the nfs cluster; goes through all the exports and starting from 1 it returns the next available integer that can be used. so in case of export creation failure, nothing is committed in the cluster and thus next time it will start from `(last export_id in cluster)+1`

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: and on top of it you have this line:https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L637

so now let's say have a export block like and you already have an export in cluster with export_id 1: 
```
  {
    "path": "dir1",
    "cluster_id": "my-nfs",
    "pseudo": "/ceph",
    "access_type": "RW",
    "squash": "none",
    "security_label": true,
    "protocols": [
      4
    ],
    "transports": [
      "TCP"
    ],
    "fsal": {
      "name": "CEPH",
      "user_id": "foo",  -----> invalid
      "fs_name": "myfs"
    },
    "clients": []
  }
```


, the output json would be:
```
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.test.2'",
    "state": "error"
  }
]
```

the 2 in suffix is the export_id.

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: the point is anyone who is using this should be knowing/must know about this behaviour that we provide export_id to the export(if not provided) depending on the exports in the cluster. This is also a good time to add this in the NFS docs as a note

**Comment by dparmar18** [comment]
last push was just a rebase

File: qa/tasks/cephfs/test_nfs.py
Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:935]: Yeh, I understand the code part. My question was more of related to the export_id returned in the JSON on failure. I understand that the export_id isn't committed. Maybe returning the export_id is fine since its unused but I hope users do not get confused with this behaviour.

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?

If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.

Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:935]: > removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?
> 
> If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.

You mean pseudo_path, isn't it? That sounds  fine to me.

Comment (vshankar) [reply qa/tasks/cephfs/test_nfs.py:935]: @dparmar18 Let's have this change since it does not confuses users.

Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py:935]: > > removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?
> > If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.
> 
> You mean pseudo_path, isn't it? That sounds fine to me.

yes pseudo is pseudo path, we don't have anything like pseudo_path

**Comment by dparmar18** [comment]
last push: no change to code; just some trivial changes to commit msg and description

**Comment by dparmar18** [comment]
last push: since we're now returning pseudo path instead of export_ids, there was a doc string that needed s/export_id/pseudo path https://github.com/ceph/ceph/compare/3a95a473be1a0bc2bb76989b0ad592894e31a8de..2f47273bae507cfdcbbf2fcc89ab57b6ad596230

**Comment by vshankar** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
LGTM

**Comment by vshankar** [comment]
unrelated fs:nfs failure - https://pulpito.ceph.com/vshankar-2023-10-20_05:01:11-fs:nfs-wip-vshankar-testing-20231013.093215-testing-default-smithi/

none of the tests ran, so I'm delaying merging this till we have it running. @dparmar18

**Comment by vshankar** [comment]
Test run in ~2h - wip-vshankar-testing-20231127.102914

**Comment by dparmar18** [comment]
> @dparmar18 pls rebase and push.

done

File: qa/tasks/cephfs/test_nfs.py
Comment (dparmar18) [reply qa/tasks/cephfs/test_nfs.py]: https://github.com/ceph/ceph/compare/52a286e49e0ac0e6f6c7dc18aea60b405f9db90b..e1ebfc023a929689c2090bafcde38861817f9f37

added assertion for errno and removed the string comparison.

**Comment by dparmar18** [comment]
@vshankar are there anymore changes required?

**Comment by vshankar** [comment]
> @vshankar are there anymore changes required?

Please tag me in the PR once its ready for (re)review.

**Comment by dparmar18** [comment]
> > @vshankar are there anymore changes required?
> 
> Please tag me in the PR once its ready for (re)review.

@vshankar it's ready for review

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#29-Jan-2024

**Pull request merged**
